platform/engineering/design-docs/2021-01-13-web-component-library.md

```diff
@@ -0,0 +1,221 @@
+
+1. each Component is initially defined as a wrapper
+1. one of many Components appears in the DOM
+1. wrapper Component lazy-loads the _actual_ definition
+1. properly defined Component is rendered in the browser
+
+[Stencil has a useful article that talks about their lazy loading approach](https://stenciljs.com/blog/how-lazy-loading-web-components-work).
+
+> It doesn’t matter how the element came to be rendered in the DOM, either way, Stencil knows how to lazy-load itself using web-standards and native browser APIs. And by deferring most of the logic to the browser, rather than generating more JavaScript runtime, the components themselves have an optimal file size.
+
+#### Migration to Web Components
+
+Once a new release is made for the Web Component library, we will [configure a linting rule](https://github.com/AlexMost/eslint-plugin-deprecate/blob/master/docs/rules/import.md) as part of the `lint:js:changed` script in `vets-website` to throw an error if a React component is imported that has a released Web Component replacement. This will prevent the older React component from appearing in more places as we are actively making efforts to remove it.
+
+Next, the Design System team will reach out to a VFS team that we believe will get the gratest value by adopting the Component. We will briefly describe the new Component and how it can be used in place of the older components currently used in their codebase, and we will offer to set up a meeting to synchronously go over the Component and begin the migration. The focus will be on making the migration as smooth as possible for the team as well as bringing awareness to the new Component.
+
+When we have finished helping the team transition all of their code to use the new Component, we will announce this publicly in Slack, and move on to help a new team with the migration, repeating the steps. After the new Web Component has replaced all instances of its React/`formation`/liquid conterparts, the React component will be removed from `component-library` and a new version will be published. The associated styles will also be removed from `formation` if we are certain that this will not affect other styles on va.gov.
+
+Additionally, the Storybook stories for the React component will be replaced by a simple document that points the user to a different Storybook host that is exclusively for Web Components. See [Chromatic hosting](#chromatic-hosting) for possible future work to improve on this.
+
+This process will repeat for each Component that has React/`formation`/liquid counterparts to replace.
+
+### Code Location
+
+The code will live in the `component-library` repo on a dedicated orphan branch named `web-components`. This code will be used in `vets-website` via a dependency in `package.json`
+
+### Testing Plan
```

Comment:
How will this be tested in `vets-website`? Do we need to create any guidance? Tools?

Reply:
Some tests may need to be updated, i.e. the ones that are looking at the DOM for things specific to the React components. Some `skin-deep` tests use shallow rendering and may have a test that expects a `CollapsiblePanel` node to be rendered, for example. Other components may have specific classes that a test might look for, but with the Web Component implementation these classes may no longer be there or be within the shadow DOM.

I don't think we need to provide any tools specific to `vets-website`. If we find that there are common problems with a non-obvious solution related to testing, then anytime we make an announcement about a new WC I think we should include a blurb about that, like:

> Q: I'm getting [SPECIFC TEST ERROR], how do I fix it?
> 
> A: This has happened with other teams/components. We recommend XYZ.

Reply:
I guess what I'm asking about is how web components will interact with the existing e2e and unit tests. Like, how will they get loaded?

So I know we'll need to change those `skin-deep` tests, but when we interact with the web component in the unit test—like clicking on the `<va-accordion>` to ensure that the expected content is there—how will the functionality be there? We're using JSDOM and—presumably—aren't adding the WC loader to it to register the components and provide functionality. Unless I'm wrong? (I know basically nothing about testing web components.)

The tool I was asking about is more along the lines of providing the web component functionality in downstream (`vets-website`) unit tests which have to interact with the components.

If I'm still not making sense (because words are hard :sweat_smile:), please reach out in Slack and we can talk there. :+1: